### PR TITLE
add config.hpp for external usage

### PR DESCRIPTION
**Pull request by simogasp**
CMake generates a `config.hpp` with the preprocessor definitions
* It is placed in a `generated` folder
* it is installed in `include/openMVG`
* all the header/source depending on those definitions now include `openMVG/config.hpp`
* all the definitions are now in the form `OPENVMG_USE_*` or `OPENVMG_HAVE_*`

Incidentally, 
* it fixes also a bug/typo `OPENVMG_USE_SSE` --> `OPENMVG_USE_SSE`, so that it was never taken into account.
* makes c++11 default removing legacy code for `Timer` class and removing `tinythread` dependency
* also refactoring of the `OPENVMG_USE_OPENMP` connected to alicevision/openMVG#291 
* also fix the public libraries propagation connected to alicevision/openMVG#235

connected to alicevision/openMVG#192

### CODE REVIEW

**Comment by fabiencastan** [comment]
This PR is great as, it will properly propagate defines when used as 3rd party.
But with this PR defines are no more defined globally by cmake but require the include of "config.hpp".
If an include is removed somewhere, the condition will change from one place to another. So I think, we really need a way to ensure that it will generate an error at compile time if an include of "config.hpp" is missing.

Proposition:
* in config.hpp
```
// set defines as functions
#define USE_OCVSIFT() 1
#define HAVE_ALEMBIC() 0

// Add function to check defines
#define OPENMVG_IS_DEFINED(F) F() == 1
```
* in another file
```
#include <openMVG/config.hpp> ///< if the include is missing it will generate an error as OPENMVG_IS_DEFINED will be missing and the function cannot be evaluated. In opposition to the current PR implementation where USE_OCVSIFT will be considered FALSE as not defined.
int main()
{
  #if OPENMVG_IS_DEFINED(USE_OCVSIFT) ///< if there is a typo in the name like USE_OVCSIFT, it will generate an error at compile time as all defines are used as define functions.
    std::cout << "I can use OCSIFT safely." << std::endl;
  #else
    std::cout << "Without OCVSIFT" << std::endl;
  #endif
};
```

**Comment by simogasp** [comment]
sounds ok to me.
I still have to finish some changes but then i can switch to that implementation

**Comment by fabiencastan** [comment]
As we are there, could be nice to also add the prefix "OPENMVG_" to all defines that are not yet prefixed like "HAVE_OPENGV", "USE_OCVSIFT", etc. to avoid any conflict with other libraries.

**Comment by simogasp** [comment]
yes it was already on the todo list

**Comment by simogasp** [comment]
for the defines in `config.hpp` do we need some "protection" like eg

```c++
#pragma once

#ifndef USE_OCVSIFT
#define USE_OCVSIFT() 1
#else
#error "USE_OCVSIFT is already defined!"
#endif
```
This would warn if, eg, USE_OCVSIFT is defined somewhere else or in compile command line. Is it overdoing it? :-)

**Comment by fabiencastan** [comment]
It's too much IMO

**Comment by simogasp** [comment]
@fabiencastan @yann-lty 
that should be all about it.
There is still a minor quirk to fix, we should enforce the host application to use C++11 when using omvg as 3rd party. One way to go is to add
```cmake
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
in `OpenMVGConfig.cmake` otherwise I don't know how to propagate that automatically, unless adding a proper 
```cmake
set_target_properties(targets PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)
```
for each omvg target (in omvg CMakeLists).

File: src/openMVG/types.hpp [hunk 1]
```diff
@@ -17,7 +17,7 @@
 #include <set>
 #include <vector>
 
-#if defined OPENMVG_STD_UNORDERED_MAP
+#ifdef OPENMVG_STD_UNORDERED_MAP
```
Comment (fabiencastan) [inline src/openMVG/types.hpp]: `#if OPENMVG_IS_DEFINED(OPENMVG_STD_UNORDERED_MAP)`

File: src/software/localization/main_rigLocalizer.cpp [hunk 2]
```diff
@@ -26,9 +27,9 @@
 #include <chrono>
 #include <memory>
 
-#ifdef HAVE_ALEMBIC
+#if OPENMVG_IS_DEFINED(OPENMVG_HAVE_ALEMBIC)
 #include <openMVG/sfm/AlembicExporter.hpp>
-#endif // HAVE_ALEMBIC
+#endif // OPENMVG_HAVE_ALEMBIC
```
Comment (fabiencastan) [inline src/software/localization/main_rigLocalizer.cpp]: `#endif`
useless comment

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -129,30 +133,24 @@ endif()
...
+if(NOT CXX11_COMPILER)
+  message(FATAL_ERROR "The compiler does not support the CXX11 standard.")
+endif(NOT CXX11_COMPILER)
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 # ==============================================================================
 # OpenMP detection
 # ==============================================================================
+set(OPENMVG_USE_OPENMP 0)
```
Comment (simogasp) [inline src/CMakeLists.txt]: Nope, the if checks for `**OpenMVG**_USE_OPENMP` and not for `**OPENMVG**`_USE_OPENMP, so it does not bug. That being said I'm not a fun of this notation (capital, non-capital) either as it can easily mess up things. I used the same preprocessor definition that there was before, I can rename it to `OPENMVG_HAVE_OPENMP`, it's maybe better.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -129,30 +133,24 @@ endif()
...
 IF (OpenMVG_USE_OPENMP)
   FIND_PACKAGE(OpenMP)
   IF (OPENMP_FOUND)
```
Comment (simogasp) [inline src/CMakeLists.txt]: Yes and indeed the very same it is declared at the beginning, so the ones inside the for can be removed. 
Also, but this could be another PR, I'd really prefer that if an option set by the user is not satisfied the cmake will fatal error rather than silently fallback to `OFF`. What do u think?

File: src/openMVG/types.hpp (hunk 1 above)
Comment (simogasp) [inline src/openMVG/types.hpp]: Yes but `OPENMVG_STD_UNORDERED_MAP` is not currently defined anywhere, I think it comes from the time c++11 was not mandatory. In that case it can be even removed and `unordered_map`used by default. Unless it is left as test code. In that case I can add it as an option I guess

File: src/CMakeLists.txt (hunk 4 above)
Comment (fabiencastan) [inline src/CMakeLists.txt]: Yes, I was looking for that too. I already had troubles due to these silent fallbacks.

It's really sad that cmake does not allow OPTIONs with 3 status or a way to know if the user as explicitly set this option.

But yes, we will do that in another PR if we found a solution !

File: src/openMVG/types.hpp (hunk 1 above)
Comment (fabiencastan) [inline src/openMVG/types.hpp]: Ah ok, so forget the comment.

**Comment by simogasp** [comment]
i'll fix the other stuff asap

File: src/CMakeLists.txt (hunk 3 above)
Comment (simogasp) [inline src/CMakeLists.txt]: @fabiencastan so shall I rename it `OPENMVG_HAVE_OPENMP`?

Comment (simogasp) [inline src/CMakeLists.txt]: there is also USE_OCVSIFT ...

Comment (fabiencastan) [inline src/CMakeLists.txt]: Ok, we will change that later in another PR.
